Wikipedia/UI-V5/WMFAppViewController.m

```diff
@@ -142,14 +155,24 @@ - (WMFArticleListCollectionViewController*)recentArticlesViewController {
 
 #pragma mark - UIViewController
 
-- (void)viewDidLoad {
-    [super viewDidLoad];
-    [self showSplashView];
-
-    [self runDataMigrationIfNeededWithCompletion:^{
-        [self hideSplashViewAnimated:YES];
-        [self loadMainUI];
-    }];
+- (void)viewDidAppear:(BOOL)animated {
+    [super viewDidAppear:animated];
+
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        [self showSplashView];
+
+        [self runDataMigrationIfNeededWithCompletion:^{
+            [self loadMainUI];
+            if ([self shouldShowOnboarding]) {
+                [self showOnboardingWithCompletion:^{
+                    [self hideSplashViewAnimated:NO];
+                }];
+            } else {
+                [self hideSplashViewAnimated:YES];
```

Comment:
To simplify the high level flow, I would take the approach of `runDataMigrationIfNeededWithCompletion:`method.

If you do `showOnboardingIfNeededWithCompletion` than the if/else statement becomes unnecessary here. It just makes the code more readable here (and pushes the if/else inside of the `showOnboardingIfNeededWithCompletion method`)